chalk-rust-ir/src/lib.rs

```diff
@@ -90,6 +90,44 @@ pub struct StructFlags {
     pub fundamental: bool,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FnDefDatumBound<I: Interner> {
+    pub params: Vec<Ty<I>>,
+    pub returns: Vec<Ty<I>>,
+    pub where_clauses: Vec<QuantifiedWhereClause<I>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FnDefFlags {
+    pub upstream: bool,
+    pub fundamental: bool,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FnDefDatum<I: Interner> {
+    pub binders: Binders<FnDefDatumBound<I>>,
+    pub id: FnDefId<I>,
+    pub flags: FnDefFlags,
+}
+
+impl<I: Interner> FnDefDatum<I> {
+    pub fn name(&self, interner: &I) -> TypeName<I> {
+        self.id.cast(interner)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ClosureDatum<I: Interner> {
+    //pub binders: Binders<StructDatumBound<I>>,
+    pub id: ClosureId<I>,
+}
+
+impl<I: Interner> ClosureDatum<I> {
+    pub fn name(&self, interner: &I) -> TypeName<I> {
+        self.id.cast(interner)
+    }
+}
+
```

Comment:
Why are these here, but the others are in `chalk-ir`? Plus there's duplicate structs like `FnDefFlags`/`FnFlags`

Reply:
The current implementation is pretty much just a duplication and slight modification of `Struct` as a starting point.  For example, [`chalk_ast::StructFlags`](http://rust-lang.github.io/chalk/chalk_parse/ast/struct.StructFlags.html) and [`chalk_rust_ir::StructFlags`](http://rust-lang.github.io/chalk/chalk_rust_ir/struct.StructFlags.html) turned into `FnDefFlags` and `FnFlags`.  And `FnDefDatum` is here because [`StructDatum`](http://rust-lang.github.io/chalk/chalk_rust_ir/struct.StructDatum.html) is also here!

I'm oscillating between thinking I understand the goal and the pieces I'm working with and thinking I don't understand it :smile: .  I'll go back to the basics and see if I can get a better understanding of the pieces involved here!  The review helped me realize that my working understanding of [`Field`](http://rust-lang.github.io/chalk/chalk_parse/ast/struct.Field.html) was off base, for example, and perhaps I've missed other important details!